kafka/src/main/java/com/metamx/tranquility/kafka/KafkaConsumer.java

```diff
@@ -0,0 +1,273 @@
+  private final int numThreads;
+  private final int commitMillis;
+  private final WriterController writerController;
+
+  private Map<String, MessageCounters> previousMessageCounters = new HashMap<>();
+
+  public KafkaConsumer(
+      final TranquilityKafkaConfig globalConfig,
+      final Properties kafkaProperties,
+      final Map<String, DataSourceConfig<TranquilityKafkaConfig>> dataSourceConfigs,
+      final WriterController writerController
+  )
+  {
+    this.consumerConnector = getConsumerConnector(kafkaProperties);
+    this.topicFilter = new Whitelist(buildTopicFilter(dataSourceConfigs));
+
+    log.info("Kafka topic filter [%s]", this.topicFilter);
+
+    int defaultNumThreads = Math.max(1, Runtime.getRuntime().availableProcessors() - 1);
+    this.numThreads = globalConfig.getConsumerNumThreads() > 0
+                      ? globalConfig.getConsumerNumThreads()
+                      : defaultNumThreads;
+
+    this.commitMillis = globalConfig.getCommitPeriodMillis();
+    this.writerController = writerController;
+    this.consumerExec = Execs.multiThreaded(numThreads, "KafkaConsumer-%d");
+    this.commitThread = new Thread(createCommitRunnable());
+    this.commitThread.setName("KafkaConsumer-CommitThread");
+    this.commitThread.setDaemon(true);
+  }
+
+  public void start()
+  {
+    commitThread.start();
+    startConsumers();
+  }
+
+  public void stop()
+  {
+    if (shutdown.compareAndSet(false, true)) {
+      log.info("Shutting down - attempting to flush buffers and commit final offsets");
+
+      try {
+        commitLock.writeLock().lockInterruptibly(); // prevent Kafka from consuming any more events
+        try {
+          writerController.flushAll(); // try to flush the remaining events to Druid
+          writerController.stop();
+          consumerConnector.commitOffsets(); // update commit offset
+        }
+        finally {
+          commitLock.writeLock().unlock();
+          consumerConnector.shutdown();
+          commitThread.interrupt();
+          consumerExec.shutdownNow();
+        }
+      }
+      catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        Throwables.propagate(e);
+      }
+
+      log.info("Finished clean shutdown.");
+    }
+  }
+
+  public void join() throws InterruptedException
+  {
+    commitThread.join();
+  }
+
+  void commit() throws InterruptedException
+  {
+    commitLock.writeLock().lockInterruptibly();
+    try {
+      final long flushStartTime = System.currentTimeMillis();
+      final Map<String, MessageCounters> messageCounters = writerController.flushAll(); // blocks until complete
+
+      final long commitStartTime = System.currentTimeMillis();
+      consumerConnector.commitOffsets();
+
+      final long finishedTime = System.currentTimeMillis();
+      Map<String, MessageCounters> countsSinceLastCommit = new HashMap();
+      for (Map.Entry<String, MessageCounters> entry : messageCounters.entrySet()) {
```

Comment:
At some point we should add some metrics collection ability but that does not have to be done in this PR